### PR TITLE
M4 - Impl mina Merkle hasher and merger with Poseidon hash algorithm

### PR DESCRIPTION
**Pull request by hanabi1224**
**Summary of changes**
Impl mina Merkle hasher and merger with Poseidon hash algorithm. 

Note: It does not define `Account` data structure, which is already defined in https://github.com/ChainSafe/mina-rs/pull/183
We just need to implement `mina_hasher::Hashable` for that `Account` data structure later.

Don't merge before below upstream PRs are finished:
- https://github.com/o1-labs/proof-systems/pull/479
- https://github.com/o1-labs/proof-systems/pull/462
- https://github.com/o1-labs/proof-systems/pull/460
- https://github.com/o1-labs/proof-systems/pull/463

Changes introduced in this pull request:
- impl `MinaPoseidonMerkleMerger`
- impl `MinaPoseidonMerkleHasher`
- basic unit tests
- parity tests with genesis ledger for the merger

**Reference issue to close (if applicable)**

Closes https://github.com/ChainSafe/mina-rs/issues/155 https://github.com/ChainSafe/mina-rs/issues/156

**Other information and links**

### CODE REVIEW

File: merkle/tests/tree_poseidon.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        assert_eq!(h2, merged);
+    }
+
+    #[cfg(not(target_arch = "wasm32"))]
+    #[test]
+    fn genesis_ledger_parity_test() {
+        use rocksdb::*;
+
+        let db =
+            DB::open_for_read_only(&Options::default(), "../ledger/test-data/genesis_ledger_6a887ea130e53b06380a9ab27b327468d28d4ce47515a0cc59759d4a3912f0ef/", true).unwrap();
```
Comment (willemolding) [inline merkle/tests/tree_poseidon.rs:116]: I think ideally our DbBackedGenesisLedger should be useable here. This is a lot of messy database logic to include in a test

File: merkle/src/merger_poseidon.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        }
+    }
+}
+
+fn merge_poseidon_hash(hashes: [Option<Fp>; MINA_POSEIDON_MERKLE_DEGREE], height: u32) -> Fp {
+    static HASHER_POOL: OnceCell<
+        SpinLockObjectPool<PoseidonHasherLegacy<MinaPoseidonMerkleTreeNonLeafNode>>,
+    > = OnceCell::new();
+    // Not calling reset here because `hasher.init` is called after `pull`, which implicitly calls sponge.reset()
+    let pool = HASHER_POOL.get_or_init(|| SpinLockObjectPool::new(|| create_legacy(0), |_| ()));
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply merkle/src/merger_poseidon.rs:55]: The creation of inner sponge is very expensive, with pooling we get ~2-3x perf boost, here's a benchmark

https://github.com/o1-labs/proof-systems/pull/462

```
Benchmarking legacy hasher without pooling: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 7.2s, enable flat sampling, or reduce sample count to 50.
legacy hasher without pooling                                                                             
                        time:   [1.4237 ms 1.4293 ms 1.4361 ms]
                        change: [+0.5134% +1.6572% +2.8203%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe

legacy hasher with pooling                                                                            
                        time:   [663.70 us 675.05 us 688.91 us]
                        change: [-2.4651% -0.5598% +1.6008%] (p = 0.59 > 0.05)
                        No change in performance detected.
Found 16 outliers among 100 measurements (16.00%)
  4 (4.00%) high mild
  12 (12.00%) high severe

Benchmarking kimchi hasher without pooling: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.7s, enable flat sampling, or reduce sample count to 60.
kimchi hasher without pooling                                                                             
                        time:   [1.0998 ms 1.1075 ms 1.1160 ms]
                        change: [-1.3130% +0.2490% +1.6556%] (p = 0.74 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) high mild
  5 (5.00%) high severe

kimchi hasher with pooling                                                                            
                        time:   [651.03 us 660.51 us 672.22 us]
                        change: [-0.7620% +1.3435% +3.5348%] (p = 0.23 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) high mild
  7 (7.00%) high severe
```

File: merkle/src/merger_poseidon.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                        print_string empty
+                */
+                "14604874247461951431777712543359658136906556694369689076707549712589474483312",
+                10,
+            )
+            .expect("Failed to parse BigUint");
+            let big256: BigInteger256 = big
+                .try_into()
+                .expect("Failed to convert BigUint to BigInteger256");
+            big256.into()
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply merkle/src/merger_poseidon.rs:101]: Agreed. I will put radix 10 representation in comment only

File: merkle/src/tree_impl.rs [hunk 4]
```diff
@@ -36,8 +36,8 @@ where
 
     /// Creates a new instance of [MinaMerkleTree] with estimated capacity of leaves
     pub fn with_capacity(capacity: usize) -> Self {
-        let protential_depth = calculate_depth(capacity);
-        let protential_node_count = calculate_node_count(protential_depth);
+        let protential_height = calculate_height(capacity);
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply merkle/src/tree_impl.rs:39]: Ah, good catch!

File: merkle/tests/tree_poseidon.rs (hunk 1 above)
Comment (hanabi1224) [reply merkle/tests/tree_poseidon.rs:116]: Yeah, I will leave a TODO here

File: merkle/src/merger_poseidon.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hanabi1224) [reply merkle/src/merger_poseidon.rs:101]: Regarding `lazy_static` and `once_cell`, I think the latter should be prefered as it has made into nightly std lib

[the remark this answers is not in the document]
Comment (hanabi1224) [reply merkle/src/merger_poseidon.rs]: Fixed

File: merkle/tests/tree_poseidon.rs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            if height > max_height {
+                max_height = height;
+            }
+            if let Some(vec) = height_2_nodes.get_mut(&height) {
+                vec.push(hash);
+            } else {
+                height_2_nodes.insert(height, vec![hash]);
+            }
+        }
+        let mut assert_hit = false;
```
Comment (creativcoder) [inline merkle/tests/tree_poseidon.rs:140]: Should this line be within the loop at the start?

Comment (hanabi1224) [reply merkle/tests/tree_poseidon.rs:140]: No this one is to make sure the ledger is not empty and the assertion happens

**Review by creativcoder — APPROVED** [review]
Great progress with this! :white_check_mark:

File: merkle/src/merger_poseidon.rs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+use super::*;
+use crate::prefixes::*;
+use ark_ff::{BigInteger256, FromBytes};
+use lockfree_object_pool::{SpinLockObjectPool, SpinLockReusable};
+use mina_hasher::{create_legacy, Fp, Hashable, Hasher, PoseidonHasherLegacy, ROInput};
+use once_cell::sync::OnceCell;
+
+/// Merger for mina binary merkle tree that uses poseidon hash
+/// with mina specific domain string calculated from node height
+pub struct MinaPoseidonMerkleMerger {}
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply merkle/src/merger_poseidon.rs:13]: fixed

File: merkle/src/merger_poseidon.rs [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    > = OnceCell::new();
+    // Not calling reset here because `hasher.init` is called after `pull`, which implicitly calls sponge.reset()
+    let pool = HASHER_POOL.get_or_init(|| SpinLockObjectPool::new(|| create_legacy(0), |_| ()));
+    let mut hasher = pool.pull();
+    merge_poseidon_hash_with_hasher(&mut hasher, hashes, height)
+}
+
+fn merge_poseidon_hash_with_hasher(
+    hasher: &mut SpinLockReusable<PoseidonHasherLegacy<MinaPoseidonMerkleTreeNonLeafNode>>,
+    hashes: [Option<Fp>; MINA_POSEIDON_MERKLE_DEGREE],
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply merkle/src/merger_poseidon.rs]: Fixed, that's good idea!

**Review by willemolding — APPROVED** [review]

**Pull request merged**